src/backend/pagemap.h

```diff
@@ -40,19 +40,49 @@ namespace snmalloc
      */
     T* body{const_cast<T*>(&default_value)};
 
-    /**
-     * The representation of the pagemap, but nullptr if it has not been
-     * initialised.  Used to combine init checking and lookup.
-     */
-    T* body_opt{nullptr};
-
     /**
      * If `has_bounds` is set, then these should contain the
      * bounds of the heap that is being managed by this pagemap.
      */
     address_t base{0};
     size_t size{0};
 
+    /**
+     * Constant for index mask
+     */
+    static constexpr size_t INDEX_MASK =
+      ~(bits::next_pow2_const(sizeof(T)) - 1);
+
+    /**
+     * Initialisation trick requires power of 2 for the masking to be collapsed
+     * into an existing instruction.
+     */
+    static_assert(
+      bits::next_pow2_const(sizeof(T)) == sizeof(T),
+      "T's size must be a power of 2");
+
+    /**
+     * The representation the mask of addresses this mask handles.
+     * This is used so it works before properly initialised. By initialising
+     * with zero, all requests will be handled by the default value until
+     * initialisation.
+     */
+    size_t index_mask{0};
+
+    /**
+     * Optimised way to get index when not necessarily initialised.
+     *
+     * The best path had a shift and mask.  By making the mask value a global
+     * we can collapse the init check into the masking.
```

Comment:
```suggestion
     * While C/C++ conventionally encodes an optional T as a T*,
     * this requires loading the T* and branching on zero.  Instead,
     * because what we're really after is either the default_value
     * or indexing into an array, we initialize our pointer to the array
     * to point to the default_value and swap it out for the real
     * backing store upon initialization.  Initially, the mask value here
     * is zero, so all accesses are crushed to the zeroth element,
     * but later the mask will allow access to the whole backing store.
     * Thus, we avoid a branch in favor of a bitwise and ALU op.
```

It doesn't _look_ like you're combining into an existing mask instruction so much as avoiding a branch, so maybe something like this instead?  (I could be missing something!)

Reply:
The code originally (in the dissassembly) had a shift and mask to calculate the aligned address.  The change makes that particular mask a global, so we aren't increasing the instruction stream by much.  But I will add commentary similar to what you proposed with a few addtions.